WIP: Redesign functionality of GLSLcompiler class

Szymon-Rajczakowski opened the pull request:
## Description

Redesign functionality of GLSLcompiler class. Change name of this class to ShaderCompiler. Add support for HLSL, GLSL and SPV files. 



## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

Commit:
Add hlsl_raytacing_basic sample

Commit:
Add hlsl_raytacing_basic sample

Commit:
Update shaders
Update README
fix validation layers errors

Commit:
Add missing flags in update_render_pass_flags function, they should be the same as in
setup_render_pass

SaschaWillems commented:
Not sure if this should be a separate sample. Imo it would be better if this was an addition to the existing basic raytracing sample, e.g. maybe a compile time define so people can toggle between GLSL and HLSL.

Commit:
Add define HLSL_SHADER to raytracing_basic
Remove hlsl_raytracing_basic

asuessenbach commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -62,6 +62,31 @@ RaytracingBasic::~RaytracingBasic()
 	}
 }
 
+VkPipelineShaderStageCreateInfo RaytracingBasic::load_spirv_shader(const std::string &filename, VkShaderStageFlagBits stage)
```
Maybe, instead of introducing a sample-level function `load_spirv_shader`, you could extent the `load_shader` function in vk_common.cpp to also handle files with `.spv` exension by just not compiling them?
You could also easily make it a run-time switch instead of compile-time.

Szymon-Rajczakowski replied:
The run-time switch sounds good, nevertheless @SaschaWillems suggested defining compile time in a comment above. What should be the preferred solution for switching shaders in load_shader, compile-time switch or runtime switch?

SaschaWillems replied:
I'm happy with both. Feel free to implement it the way you prefer.

asuessenbach commented on shaders/khr_ray_tracing_basic/hlsl_closesthit.rchit:
```diff
@@ -0,0 +1,33 @@
+[shader("closesthit")]
+void main(inout Payload payload, in Attribute attribs)
+{
+    const float3 barycentrics = float3(1.0 - attribs.bary.x - attribs.bary.y, attribs.bary.x, attribs.bary.y);
```
I don't know much about HLSL, but wouldn't it be better to have `1.0f` here, instead of `1.0`?

Commit:
- Change glsl_compiler class name
to shader_compiler
- Add ShaderSourceLanguage enum
- Add src_language parameter to load_shader for
support different shader languages
- Change headers and few names in existing samples

asuessenbach commented on framework/common/vk_common.cpp:
```diff
@@ -356,25 +356,36 @@ int32_t get_bits_per_pixel(VkFormat format)
+			return VK_NULL_HANDLE;
+		}
+	}
+	if (src_language == vkb::ShaderSourceLanguage::VK_HLSL)
```
Should probably be `else if` here

Szymon-Rajczakowski replied:
I have refactored this method and removed unnecessary logic. I will resolve all conversation from old vk_common.cpp. I would appreciate it if you can find the time to review the latest changes

asuessenbach commented on framework/common/vk_common.cpp:
```diff
@@ -356,25 +356,36 @@ int32_t get_bits_per_pixel(VkFormat format)
+		{
+			LOGE("Failed to compile shader, Error: {}", info_log.c_str());
+			return VK_NULL_HANDLE;
+		}
```
Where's the difference between the two cases here, GLSL vs. HLSL?

asuessenbach commented on framework/common/vk_common.cpp:
```diff
@@ -383,6 +394,11 @@ VkShaderModule load_shader(const std::string &filename, VkDevice device, VkShade
 	module_create_info.codeSize = spirv.size() * sizeof(uint32_t);
 	module_create_info.pCode    = spirv.data();
```
These two assignments are only needed when not VK_SPV.

asuessenbach commented on framework/shader_compiler.cpp:
```diff
@@ -81,48 +81,58 @@ inline EShLanguage FindShaderLanguage(VkShaderStageFlagBits stage)
 
+	if (src_language == vkb::ShaderSourceLanguage::VK_HLSL)
+	{
+		messages = static_cast<EShMessages>(EShMsgReadHlsl | EShMsgDefault | EShMsgVulkanRules | EShMsgSpvRules);
```
Maybe a bit more clear, that you just add some hsls support:
`messages = static_cast<EShMessages>(messages | EShMsgReadHlsl);`

asuessenbach commented on framework/shader_compiler.h:
```diff
@@ -56,17 +56,18 @@ class GLSLCompiler
 	                      std::vector<std::uint32_t> &spirv,
-	                      std::string &               info_log);
+	                      std::string                &info_log,
+	                      vkb::ShaderSourceLanguage   src_language = vkb::ShaderSourceLanguage::VK_GLSL);
```
I think, the logical position of the `src_language` argument would be between the `stage` and the `shader_source`. Even though, it would need adjustments all over the samples and the framework.
Or maybe even as a template argument to `compile_to_spirv`?
Any opinion on that?

Szymon-Rajczakowski replied:
I've overloaded the compile_to_spirv method. Do you think this is a good solution?

asuessenbach commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -603,21 +578,16 @@ void RaytracingBasic::create_ray_tracing_pipeline()
-#ifdef HLSL_SHADER
-		shader_stages.push_back(load_spirv_shader("khr_ray_tracing_basic/hlsl_raygen.rgen.spv", VK_SHADER_STAGE_RAYGEN_BIT_KHR));
-#else
 		shader_stages.push_back(load_shader("khr_ray_tracing_basic/raygen.rgen", VK_SHADER_STAGE_RAYGEN_BIT_KHR));
```
No differentiation between GLSL and HLSL anymore?

Szymon-Rajczakowski replied:
I have made a change that will allow you to switch between GLSL shaders and SPV shaders using the UI. Please take a look at the latest changes

asuessenbach commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -630,11 +600,7 @@ void RaytracingBasic::create_ray_tracing_pipeline()
-#ifdef HLSL_SHADER
-		shader_stages.push_back(load_spirv_shader("khr_ray_tracing_basic/hlsl_miss.rmiss.spv", VK_SHADER_STAGE_MISS_BIT_KHR));
-#else
 		shader_stages.push_back(load_shader("khr_ray_tracing_basic/miss.rmiss", VK_SHADER_STAGE_MISS_BIT_KHR));
```
As above.

asuessenbach commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -647,11 +613,7 @@ void RaytracingBasic::create_ray_tracing_pipeline()
-#else
-		shader_stages.push_back(load_shader("khr_ray_tracing_basic/closesthit.rchit", VK_SHADER_STAGE_CLOSEST_HIT_BIT_KHR));
-#endif
+		shader_stages.push_back(load_shader("khr_ray_tracing_basic/hlsl_closesthit.rchit.spv", VK_SHADER_STAGE_CLOSEST_HIT_BIT_KHR, vkb::VK_SPV));
```
As above.

asuessenbach reviewed (commented):
I tried to run hlsl_shaders, but get the following error:
```
[error] [framework\common\vk_common.cpp:386] Failed to compile shader, Error: ERROR: #version: ES shaders for SPIR-V require version 310 or higher
ERROR: :20: '' :  syntax error, unexpected LEFT_BRACKET
ERROR: 2 compilation errors.  No code generated.
```
Would I need to update anything on my end?

Besides that: is there any means to switch between HSLS and GLSL now?

Commit:
-Overloading the compile_to_spirv method
to keep the order of arguments more meaningful
- Add switch in UI to changing shaders language in
raytracing_basic sample

Commit:
Rebase and resolve commit problems

Commit:
Fixed quality of code and copyrights

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=655) <br/>All committers have signed the CLA.

lorna-mcneill-mobica committed:
Fix build breaks

asuessenbach commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -577,18 +577,20 @@ void RaytracingBasic::create_ray_tracing_pipeline()
 
-	// Ray tracing shaders require SPIR-V 1.4, so we need to set the appropriate target environment for the glslang compiler
-	vkb::GLSLCompiler::set_target_environment(glslang::EShTargetSpv, glslang::EShTargetSpv_1_4);
+	// Ray tracing shaders require SPIR-V 1.4, so we need to set the appropriate target envinroment for the glslang compiler
```
introduced typo: envinroment

asuessenbach commented on framework/api_vulkan_sample.cpp:
```diff
@@ -427,12 +427,12 @@ void ApiVulkanSample::create_pipeline_cache()
 }
 
-VkPipelineShaderStageCreateInfo ApiVulkanSample::load_shader(const std::string &file, VkShaderStageFlagBits stage)
+VkPipelineShaderStageCreateInfo ApiVulkanSample::load_shader(const std::string &file, VkShaderStageFlagBits stage, vkb::ShaderSourceLanguage src_language)
```
Would you mind to also extend HPPApiVulkanSample::load_shader accordingly?

asuessenbach commented on framework/api_vulkan_sample.cpp:
```diff
@@ -828,17 +828,17 @@ void ApiVulkanSample::update_render_pass_flags(uint32_t flags)
 	dependencies[0].dstSubpass      = 0;
 	dependencies[0].srcStageMask    = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT;
-	dependencies[0].dstStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+	dependencies[0].dstStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT | VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
```
Could you please adjust the corresponding values in HPPApiVulkanSample::setup_render_pass?

lorna-mcneill-mobica replied:
@asuessenbach 

You mean HPPApiVulkanSample::update_render_pass_flags?

I see that on KhronosGroup:main, the HPP and non-HPP versions of update_render_pass_flags have diverged, eg 

dependencies[1].srcStageMask    = vk::PipelineStageFlagBits::eColorAttachmentOutput | vk::PipelineStageFlagBits::eEarlyFragmentTests | vk::PipelineStageFlagBits::eLateFragmentTests;

versus 

dependencies[1].srcStageMask    = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;

(not sure whether there are other differences)

I guess those shoule be aligned to whichever is correct too?

asuessenbach replied:
> I guess those should be aligned to whichever is correct too?

That would be perfect, yes!

lorna-mcneill-mobica replied:
@asuessenbach 
I had a look at the change that caused that diversion. It's quite a significant change made to fix various synchronisation errors reported by validation.
If it's Ok with you I'd prefer to keep that "re-alignment" separate from Szymon's change so as not to muddy the waters?

(In that case I suppose I would do that as a separate PR)

asuessenbach replied:
Sure, I'm fine with that.
We should just make sure to not miss that re-alignment.

asuessenbach commented on framework/common/vk_common.cpp:
```diff
@@ -356,25 +356,28 @@ int32_t get_bits_per_pixel(VkFormat format)
 }
 
-VkShaderModule load_shader(const std::string &filename, VkDevice device, VkShaderStageFlagBits stage)
+VkShaderModule load_shader(const std::string &filename, VkDevice device, VkShaderStageFlagBits stage, vkb::ShaderSourceLanguage src_language)
```
Would you mind to also extend `load_shader` in hpp_vk_common.h?

asuessenbach commented on framework/shader_compiler.h:
```diff
@@ -54,19 +54,28 @@ class GLSLCompiler
+	                      const std::string          &entry_point,
+	                      const ShaderVariant        &shader_variant,
+	                      std::vector<std::uint32_t> &spirv,
+	                      std::string                &info_log);
```
Instead of having two flavours of `compile_to_spirv` I would prefer to have just one. Maybe with some default value for the `src_language`.
And please extend the HPPShaderCompiler accordingly.

lorna-mcneill-mobica committed:
Address review comments and fix copyright check issue

fix typo
bring HPP load_shader() and update_render_pass_flags() in line with non-hpp version
use a single compile_to_spirv() with default source language rather than have 2 flavours

asuessenbach commented on .copyrightignore:
Does this file really belong to your PR?

lorna-mcneill-mobica replied:
@asuessenbach 
Yeah, I updated this to tell the copyright checker to ignore .spv files (since Szymon added a number of binary .spv files that we do not want to be copyright checked and these were causing a failure in the CI checks)

asuessenbach commented on scripts/copyright.py:
Does this file really belong to your PR?

lorna-mcneill-mobica replied:
Ack, good catch - this was a temporary local change to track down the files causing the copyright checker to fail, I will back out the change!

lorna-mcneill-mobica committed:
back out unintentional change to copyright checking script

lorna-mcneill-mobica committed:
Fix copyright and clang-format issues

lorna-mcneill-mobica committed:
Fixes from clang-format 15.07

SaschaWillems commented on samples/extensions/raytracing_basic/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2019-2021, Sascha Willems
```
I don't think it's okay to replace the existing copyright. If you add to a file, please append your copyright instead.

SaschaWillems commented on framework/shader_compiler.h:
```diff
@@ -33,7 +33,7 @@ namespace vkb
 {
 /// Helper class to generate SPIRV code from GLSL source
```
With the compiler now being able to support other languages, the comments need to be changed too. They still refer to GLSL only.

SaschaWillems replied:
That comment still needs to be updated.

Seweryn-Zielas-Mobica replied:
Updated

SaschaWillems commented on framework/shader_compiler.h:
```diff
@@ -54,19 +54,21 @@ class GLSLCompiler
 
 	/**
-	 * @brief Compiles GLSL to SPIRV code
+	 * @brief Compiles GLSL/HLSL to SPIRV code
```
Is this correct? If so, what tool is used to compile HLSL tp SPIR-V (dxc)? Or does it simply load offline compiled SPV? If it's the latter, the comment needs to be changed.

Seweryn-Zielas-Mobica replied:
Glslang is used for compilation at runtime and for offline compilation DirectX Shader Compiler is being used.

SaschaWillems commented on samples/api/hlsl_shaders/hlsl_shaders.cpp:
```diff
@@ -410,7 +410,7 @@ void HlslShaders::prepare_pipelines()
 	// Load shaders
 	std::array<VkPipelineShaderStageCreateInfo, 2> shader_stages{};
-	shader_stages[0] = load_hlsl_shader("hlsl_shaders/hlsl_shader.vert", VK_SHADER_STAGE_VERTEX_BIT);
+	shader_stages[0] = load_shader("hlsl_shaders/hlsl_shader.vert", VK_SHADER_STAGE_VERTEX_BIT, vkb::ShaderSourceLanguage::VK_HLSL);
```
This change doesn't seem correct. This is a standalone sample that is supposed to use it's own hlsl loading function, but this line would use the one from the framework instead. The change should be reverted.

SaschaWillems replied:
This still needs to be reverted.

Seweryn-Zielas-Mobica replied:
Reverted

SaschaWillems commented on samples/api/hpp_hello_triangle/hpp_hello_triangle.cpp:
```diff
@@ -1,4 +1,4 @@
-/* Copyright (c) 2021-2023, NVIDIA CORPORATION. All rights reserved.
```
I don't think it's okay to replace the existing copyright. If you add to a file, please append your copyright instead.

SaschaWillems replied:
This is still open. This sample was done by Andreas (Nvidia).

Seweryn-Zielas-Mobica replied:
Fixed

Seweryn-Zielas-Mobica replied:
I had to revert fix, it caused CI checks for copyrights to fail

SaschaWillems commented:
I did a first quick review and added a few comments.

One thing I'm not sure about is the direction of this PR.

On one hand it seems to add HLSL support to the framework's compiler, but only through glslang, which is very very limited. And on the other hand it also adds direct loading of SPIR-V for e.g. the basic ray tracing sample.

Tbh. I think we should discuss a different approach that better fits our framework strategy.

We've been talking about HLSL support for quite some time and one option would be adding direct SPIR-V loading to the shader compiler (as a derived class maybe)  so we can add HLSL to all samples with offline compiled SPIR-V files via DXC. The goal is also to not have to change every sample but to support shader language selection at framework level instead. And imo we should move into that direction.

I appreciate the work put in this PR, but it feels a bit like a stop gap towards our actual goal and we might revert a large part of that PR in the future.

I'll wait for other opinions, and I'd happily discuss this in the next samples call :)

Patryk-Jastrzebski-Mobica commented:
Hi @SaschaWillems,
We have discussed this problem and think that the best solution will be to extend HLSL support (add raytracing support) in the glslang repository. At the moment @Dawid-Lorenz-Mobica is investigating how to add raytracing support for the HLSL language in the glslang repository.

SaschaWillems commented:
Tbh I'm not sure if that's the right approach. HLSL in glslang is very limited and the actual reference compiler is DXC. So if we want to add HLSL support we should either integrate DXC or offline compile HLSL to SPIR-V and just use that.

pj87 committed:
Post-merge compilation fixes.

Signed-off-by: Paweł Jastrzębski <pawel.jastrzebski@mobica.com>

pawel-jastrzebski-mobica committed:
Add targets to compile HLSL shaders with dxc compiler.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

SaschaWillems commented:
@Szymon-Rajczakowski-Mobica: Can you solve the merge conflict? We talked about this on our last call and would like to merge this as early as  possible, so we can start working on HLSL shaders :)

pawel-jastrzebski-mobica committed:
Merge branch 'main' into hlsl_raytracing_basic

pawel-jastrzebski-mobica committed:
Post-merge fixes.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

SaschaWillems commented:
I gave this a try and added/updated a few comments (see above) that needs to be resolved.

Several build checks also failed, e.g. Ubuntu and Android builds. These need to be fixed too.

One thing that should be changed is that right now instead of the actual human readable HLSL files, the SPV files are added to the project:

![image](https://github.com/KhronosGroup/Vulkan-Samples/assets/10283127/644f9317-d3bc-4176-8ae8-1f2fc74ff37a)

That doesn't make a lot of sense, I'd expect the hlsl source files in here.

We should also discuss folder structure. Right now (for the one sample were HLSL was added), GLSL and HLSL files are in the same folder, with the HLSL ones being prefixed. I'd prefer to separate them via folders (similar to what I do in my samples). I propose we do this with sub folders like this:

* \shaders
  * \sample
    * \glsl <-- GLSL shaders here
    * \hlsl <-- HLSL shaders and compiled SPV in here

SaschaWillems commented:
> Besides that: is there any means to switch between HSLS and GLSL now?

I was wondering the same. I don't see a way to select HLSL or GLSL. We should do that as a command line switch.

pawel-jastrzebski-mobica committed:
Update copyrights years.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

pawel-jastrzebski-mobica committed:
Change shaders paths.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

pawel-jastrzebski-mobica commented:
> > Besides that: is there any means to switch between HSLS and GLSL now?
> 
> I was wondering the same. I don't see a way to select HLSL or GLSL. We should do that as a command line switch.

@SaschaWillems: I really don't know how to command-line switch should be used in context of this change. Since the PR adds the ability compile HLSL to SPV for both off-line and on-line use-case. Please explain the functionality that could be switched using command-line.

SaschaWillems commented:
Selecting between running a sample with the GLSL based shader or the HLSL based shader.

pawel-jastrzebski-mobica committed:
Fix clang-format.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

pawel-jastrzebski-mobica committed:
Fix copyrights.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

pawel-jastrzebski-mobica committed:
Fix clang-format.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

SaschaWillems commented on docs/build.adoc:
```diff
@@ -208,13 +212,20 @@ build\windows\app\bin\Release\AMD64\vulkan_samples.exe
 * C{pp}14 Compiler
 * <<cmake-options,CMake Options>>
 * <<3d-models,3D models>>
+* Vulkan SDK
```
Can this be reverted? We don't require the SDK to build our samples.

Seweryn-Zielas-Mobica replied:
Reverted

SaschaWillems commented on docs/build.adoc:
```diff
@@ -243,11 +254,18 @@ cmake --build build/linux --config Release --target vulkan_samples -j$(nproc)
 * https://vulkan.lunarg.com/doc/sdk/latest/mac/getting_started.html[Vulkan SDK] `./install_vulkan.py`
 * <<cmake-options,CMake Options>>
 * <<3d-models,3D models>>
+* Vulkan SDK
```
Same as above.

Seweryn-Zielas-Mobica replied:
Reverted

SaschaWillems commented:
Also odd, I get three out-of-place hlsl related entries in the MSVC project browser now:

![image](https://github.com/KhronosGroup/Vulkan-Samples/assets/10283127/1328bb36-be17-4f6b-9cb5-9f72f966bf3f)

Probably left overs. Would be good if this PR did not include any HLSL shaders at all, but just the code required to load SPIR-V. We'll add the HLSL in separate PRs.

pawel-jastrzebski-mobica commented:
@SaschaWillems : do you want to remove the whole off-line shader compilation functionality from the PR?

pawel-jastrzebski-mobica committed:
Fix copyrights.

Signed-off-by: pawel-jastrzebski-mobica <pawel.jastrzebski@mobica.com>

SaschaWillems commented:
Nope, having a way to compile the HLSL shaders vie the build system is fine. Problem is where those projects are put in the project structure. They appear at the top level, but should instead be part of the sample itself.

pawel-jastrzebski-mobica committed:
Merge branch 'main' into hlsl_raytracing_basic

Seweryn-Zielas-Mobica committed:
Working build without dxc and fixed usage of dxc - missing spirv flag

Seweryn-Zielas-Mobica committed:
Modify custom command creation for DXC compiler and remove
unnecessary spv files.

Seweryn-Zielas-Mobica committed:
Merge branch 'main' into hlsl_raytracing_basic

Seweryn-Zielas-Mobica committed:
Fix build issues after merge

Seweryn-Zielas-Mobica committed:
Fix shaders names in ray tracing basic CMakeLists.txt

Seweryn-Zielas-Mobica committed:
Fix Subpass dependencies for layout transitions

Seweryn-Zielas-Mobica committed:
Modify ray_tracing_basic to use real time shader selection plugin.

Seweryn-Zielas-Mobica committed:
Fix handling shaders in hlsl_shaders and hpp_hlsl_shaders.

Seweryn-Zielas-Mobica committed:
Udpate copyrights

Seweryn-Zielas-Mobica committed:
Fix clang-format

Seweryn-Zielas-Mobica committed:
Remove Vulkan SDK dependency from build.adoc

Seweryn-Zielas-Mobica committed:
Change usage of DirectX Shader Compiler in sample_helper.cmake
from add_custom_command  to execute_process.

Seweryn-Zielas-Mobica committed:
Update comment for ShaderCompiler class and update copyright
for hpp_hello_triangle.cpp

Seweryn-Zielas-Mobica committed:
Fix copyright in hpp_hello_triangle.cpp

Seweryn-Zielas-Mobica commented:
I changed way of using dxc for offline compilation. I had problem with **add_custom_target/add_custom_command** that created targets in top level view of Visual Studio solution, I was only found way how to move it into other folders, but couldn't move it under other targets like specific samples. From what I red and understood, this is most probably not possible. 
Instead of  **add_custom_target/add_custom_command**  is used **execute_process** which seems to work as good or maybe even better than previous solution, as now code for compiling shaders seems more straightforward and less bulky.

The difference in how it works now is rather not significant, before HLSL shaders were compiled at the start of executing build command, e.g. **"cmake --build build/linux --config Release --target vulkan_samples -j$(nproc)"**
Now they are compiled after running command for generating project, e.g. "cmake -G "Unix Makefiles" -Bbuild/linux -DCMAKE_BUILD_TYPE=Release"

I also updated cmake to only compile shaders if dxc executable is found, by searching for it under **VULKAN_SDK** path.

There is still glslang used for compilation HLSL shaders at sample runtime, and rather big modification of **hpp_hello_triangle.cpp** which I'm not sure if is needed.

There is still some work left to do, mostly stylistic like naming scheme and directories/folders paths, but core functionality of this PR that is compiling HLSL shaders seems to be done.

Seweryn-Zielas-Mobica committed:
Merge branch 'main' into hlsl_raytracing_basic

SaschaWillems reviewed (changes requested):
Can you undo the addition of HLSL shaders for the ray tracing basic sample. I'm already working on adding actual HLSL shaders, and to clearly divide work this PR should only add the offline compilation part.

Seweryn-Zielas-Mobica commented:
@SaschaWillems As we discussed that we should only add offline compilation for HLSL, I've created created new PR that only do this, as it seemed easier and less misguiding when looking at this PR title.
New PR: #990

marty-johnson59 commented:
Closing this as we're going a different direction
